pydantic/utils.py

```diff
@@ -114,6 +137,79 @@ def almost_equal_floats(value_1: float, value_2: float, *, delta: float = 1e-8)
     return abs(value_1 - value_2) <= delta
 
 
+def copy_code(code: CodeType, **update: Any) -> CodeType:
+    if PY38:
+        return code.replace(**update)  # type: ignore
+    return CodeType(*[update.get(arg, getattr(code, arg)) for arg in CODE_ARGS])
+
+
+def generate_typed_init(
+    current_init: FunctionType,
+    fields: Dict[str, 'ModelField'],
+    defaults: Dict[str, Any],
+    annotations: Dict[str, Any],
+    compiled: bool,
+) -> FunctionType:
+    """
+    Generate typed signature for init
+    """
+    if not compiled:
+        # can perform this in python, however in cython this will cause "unknown opcode" on call
+        orig_init = current_init
+        orig_code = orig_init.__code__
+        new_init = FunctionType(
+            orig_code, orig_init.__globals__, closure=orig_init.__closure__, name=orig_init.__name__
+        )
+    else:
+        orig_init = getattr(current_init, '__origin_init__', current_init)
+        orig_code = orig_init.__code__
+
+        @wraps(orig_init)
+        def __init__(*args, **kwargs):  # type: ignore
+            orig_init(*args, **kwargs)
+
+        new_init = __init__  # type: ignore
+        new_init.__origin_init__ = orig_init  # type: ignore
+
+    orig_posonlycount = orig_code.co_posonlyargcount if PY38 else 0  # type: ignore
+    orig_allargscount = orig_code.co_argcount + orig_code.co_kwonlyargcount + orig_posonlycount
+    orig_argnames = orig_code.co_varnames[:orig_allargscount]
+    orig_othernames = orig_code.co_varnames[orig_allargscount:]
+
+    orig_argnames_set = set(orig_argnames)  # just for fast lookups
+    valid_fields_args = tuple([name for name in fields if name not in orig_argnames_set and name.isidentifier()])
```

Comment:
Does this generate the right signature if a field has an alias and doesn't have `allow_population_by_field_name=True`?

Reply:
Aliases are not considered at all.

And thanks for that question, I found a bug in current implementation. Model with fields declared in namespace would have this in signature:
```py
__init__(__pydantic_self__, *, a=FieldInfo(default=1, alias='A', extra={}), **data: Any) -> None
```

Can you please give me example of such model and what signature should look like?

Reply:
```python
class A(BaseModel):
    a: int = Field(alias='b')
```
It looks like your generated signature will be (ignoring any type hints):
```python
__init__(__pydantic_self__, a)
```
but if you call `__init__(a=1)` you'll get an error. I would think the signature *should * be `__init__(__pydantic_self__, b)` in this case. This is what I meant by aliases.

Reply:
Also, note that if you are using the type annotation from the field, the generated signature will actually also be wrong for any type that is not idempotent. In particular, at least currently, if you have a `SecretStr` field, you *have* to pass a `str` to the `__init__` function and *not* a `SecretStr`.  I opened an issue to fix this for `SecretStr`, but there are other types for which this is also an issue.

(Admittedly, this is also a problem in the PyCharm plugin and, in strict mode, the mypy plugin too.)

Reply:
Ok, now I see. Aliases should be considered.

Reply:
As for `SecretStr` and other custom types: I don't see any problems here since these annotations will be available only at runtime and won't bother type-checkers. Until some code checks incoming types at runtime and don't know how to deal with `pydantic`, I don't think having `pydantic` specific types in signature would be a problem. For example, `FastAPI` understands such annotations just fine:

```py
class GetUserFilters(BaseModel):
    name: Optional[constr(strip_whitespace=True, max_length=70)]
    email: Optional[SecretStr]

 
@app.get('/users')
def get_users(params=Depends(GetUserFilters)):
    ...
```
![image](https://user-images.githubusercontent.com/36469655/69902578-9d1f7100-13a0-11ea-8880-d173ced14524.png)

Reply:
I agree they won't bother type checkers, but it still feels bad for it to be generated "wrong". I think there is a tradeoff here between it faithfully representing the valid signature of the method (i.e., including a type annotation of `Any` or similar to allow for parsing), and providing useful information.

In general, I would prefer to err on the side of "correctness" (e.g., generating an `__init__` signature with `Any` as the annotation), but from a practical perspective I'm okay with it being occasionally wrong in a minor way like this if it is substantially more useful (and preferred in practice).